examples/pplm/run_pplm.py

```diff
@@ -508,6 +518,13 @@ def generate_text_pplm(
 
         pert_logits, past, pert_all_hidden = model(last, past=pert_past)
         pert_logits = pert_logits[:, -1, :] / temperature  # + SMALL_CONST
+
+        for j in set(output_so_far[0].tolist()):
+            if pert_logits[0, j] < 0:
+                pert_logits[0, j] *= repetition_penalty
```

Comment:
In the definition at the end of 4.1 in https://einstein.ai/presentations/ctrl.pdf there is no distinction between logits < 0 and > 0. I understand the reason why logits < 0 should be treated separately: dividing them by a penalty score would actually increase the value and thus increase the probability after softmax is computed, but I'd like a reference for it if possible, as in the < 0 case values may become really small. It is true that we are taking an exponential later so the differences in output between  even distant values in the negative side are negligible, but again, if there's a reference for this it would be great to add it as a comment in the code.

Reply:
Of course. I used this idea from transformers library itself as the repetition penalty was broken with GPT-2 model.

Here is the relevant PR with discussion:
https://github.com/huggingface/transformers/pull/2303

Reply:
Got it, thank you for pointing it out, I replied to that thread too: https://github.com/huggingface/transformers/pull/2303#issuecomment-572273727